### ISSUE TITLE
Suggestion: Implement SQRL

### ISSUE THREAD

**Issue by callesoe** [keybase/keybase-issues#1535]
Secure Quick Reliable Login (SQRL) seems to have serious potential for solving a lot of password hassle, including login to Keybase.

Main documentation:
https://www.grc.com/sqrl/sqrl.htm

Proposal at DigiCert Security Summit 2014:
https://www.youtube.com/watch?v=PtfFwOYxCnc

**Comment by zQueal** [comment]
This is a bit scary to me. Most of the CLI based implementations I've seen create images which need to be saved to a HDD in order to process. Even if it's highly unlikely if someone were to get ahold of this image they might be able to login to my account and subsequently access my private key.

Are there any assurances that this wont happen? (I didn't have time to read the documentation, and I assumed you were familiar with it)

**Comment by callesoe** [comment]
Disclaimer: I am no expert on SQRL, but I have been following the project for a long time.

Unless I misunderstand you, you are talking about a private key compromise scenario (the SQRL private key, not the Keybase private key), in which case the attacker has to defeat the crypto protecting it, because the SQRL key/app is never unlocked outside of signing. And you are subsequently able to replace your master key through the Identity Lock protocol.

There's explanations here with visual aids: http://sqrl.pl/guide/

**Comment by zQueal** [comment]
> the SQRL private key, not the Keybase private key

Not exactly. What I mean to say is that most scripts which generate QR codes that I've seen (about 95% of them) actually output a `jpeg` or `png` image which is then displayed. If that's true, then the server actually ends up having my password in QR form. That just kinda scares the hell out of me. If an attacker was able to gain access to Keybases' systems, and then somehow find these `png` or `jpeg` files, or was able to figure out how to create them for a given user, then my private key (which is uploaded to keybase) is now compromised.

If, however, there is no image generated, then I could see this as a viable option. Definitely.

**Comment by frank-borkin** [comment]
Zach,

That's not how SQRL works. The QR code is just a 1 time code which can be
handed out to anyone. You sign it with your private key using your SQRL app
and send it to the callback URL in the QR code. The service you're logging
into checks your signed challenge against your public key (which is how it
identifies you) and marks the 1 time code as used. There is nothing user
specific in the QR code, any code can be used to log in any user.

It's essentially the same as HTTP digest auth, OAuth or logging in via SSH
key; you just sign a challenge. The big problems are in your local SQRL app
(on your phone) handling your private key securely, especially when your
phone is stolen.

Personally, I think SQRL has a lot of issues, but the QR code isn't one of
them :)

Frank

**Comment by dtiersch** [comment]
This challenge based login is for "ordinary" platforms a very nice idea, but it won't work well with keybase - at least for the default user which has its private key hosted and uses all the functions via the website.
To decrypt the key needed for most operations you need to give a password to decrypt it, a challenge won't do as a key.

**Comment by zQueal** [comment]
@frank-borkin Hey, thanks for the input!

> To decrypt the key needed for most operations you need to give a password to decrypt it, a challenge won't do as a key.

Agreed. This project looks really promising, but to do almost any function Keybase would be authorizing itself (if this was implemented) to do what it wanted with my private key. That brings back flashbacks from [Maximum Overdrive.](http://i.imgur.com/j2s4EGJ.jpg)
